components/Starknet/modules/architecture_and_concepts/pages/Network_Architecture/messaging-mechanism.adoc

```diff
@@ -57,49 +83,55 @@ keccak256(
-. The message is then decoded into a Starknet transaction that invokes a function annotated with the `l1_handler` decorator on the target contract. Transactions like this on L2 are called *_L1 handler functions_*.
+The protocol consists of the following stages:
+
+. An L1 contract initiates a message to an L2 contract on Starknet by calling the link:https://github.com/starkware-libs/cairo-lang/blob/54d7e92a703b3b5a1e07e9389608178129946efc/src/starkware/starknet/solidity/IStarknetMessaging.sol#L13[`sendMessageToL2`] function on the Starknet Core Contract with the message parameters.
++
+The Starknet Core Contract hashes the message parameters and updates the L1->L2 message mapping to indicate that a message with this hash was indeed sent. The L1 contract records the fee that the sender paid. For more information, see xref:#l1-l2-message-fees[L1 -> L2 message fees].
+. The message is then decoded into a Starknet transaction that invokes a function annotated with the `l1_handler` decorator on the target contract. Transactions like this on L2 are called *_L1 handler transactions_*.
  .. The Starknet sequencer, upon seeing enough L1 confirmations for the transaction that sent the message, initiates the corresponding L2 transaction.
- .. The L2 transaction invokes the relevant `l1_handler`.
+// What does this mean: "upon seeing enough L1 confirmations..."? Is this talking about waiting for a batch?
+ .. The L2 transaction invokes the relevant `l1_handler` function.
 . The L1 Handler transaction that was created in the previous step is added to a proof.
-. The state update is received on the Core contract.
-. the message is cleared from the Core contract's storage. At this point, the message is handled.
+. The Core contract receives the state update.
+. The message is cleared from the Core contract's storage. At this point, the message is handled.
+
+The above flow is illustrated in the following diagram:
+#THIS IMAGE IS WRONG & MISLEADING#
 
-// The above flow is illustrated in the following diagram:
-// THIS IMAGE IS WRONG & MISLEADING AND THUS COMMENTED OUT UNTIL FIXED
-// image::l1l2.png[l1l2]
+image::l1l2.png[l1l2]
 
-An L1→L2 message consists of:
+An L1->L2 message consists of #Where is an example of an L1->L2 message?#:
 
-* The L1 sender address
-* The recipient contract address on Starknet
+* L1 sender's address
+* L2 recipient's contract address
 * Function selector
 * Calldata array
 * Message nonce
-
++
 [NOTE]
 ====
-*Message nonce*
+The message nonce is maintained on the Starknet Core contract on L1, and is #incremented# whenever a message is sent to L2. The nonce is used to avoid hash collisions between different L1 handler transactions that are #caused# [.line-through]#induced# by the same message being sent on L1 multiple times.
```

Comment:
```suggestion
The message nonce is maintained on the Starknet Core contract on L1, and is incremented whenever a message is sent to L2. The nonce is used to avoid a hash collision between different L1 handler transactions that is caused by the same message being sent on L1 multiple times.
```